src/steps/custom-object-create-or-update.ts

```diff
@@ -0,0 +1,96 @@
+/*tslint:disable:no-else-after-return*/
+
+import { BaseStep, Field, StepInterface } from '../core/base-step';
+import { Step, FieldDefinition, StepDefinition } from '../proto/cog_pb';
+
+export class CreateOrUpdateCustomObjectStep extends BaseStep implements StepInterface {
+
+  protected stepName: string = 'Create or Update a Marketo Custom Object';
+  protected stepExpression: string = 'create or update an? (?<name>.+) marketo custom object linked to lead (?<linkValue>.+)';
+  protected stepType: StepDefinition.Type = StepDefinition.Type.ACTION;
+  protected expectedFields: Field[] = [{
+    field: 'name',
+    type: FieldDefinition.Type.STRING,
+    description: "Custom Object's display or API name",
+  }, {
+    field: 'linkValue',
+    type: FieldDefinition.Type.EMAIL,
+    description: "Linked Lead's email address",
+  }, {
+    field: 'object',
+    type: FieldDefinition.Type.MAP,
+    description: 'Map of custom object data whose keys are field names.',
+  }];
+
+  async executeStep(step: Step) {
+    const stepData: any = step.getData().toJavaScript();
+    const name = stepData.name;
+    const linkValue = stepData.linkValue;
+    const object = stepData.object;
+
+    try {
+      const customObject = await this.client.getCustomObject(name);
+      // Custom Object exists validation
+      if (!customObject) {
+        return this.error('Error creating or updating %s: no such marketo custom object', [
+          name,
+        ]);
+      }
+
+      // Linked to lead validation
+      if (!customObject.result[0].relationships.some(relationship => relationship.relatedTo.name == 'Lead')) {
+        return this.error("Error creating or updating %s linked to %s: this custom object isn't linked to leads", [
+          name,
+          linkValue,
+        ]);
+      }
+
+      // DedupeField related validations
+      if (customObject.result[0].hasOwnProperty('dedupeFields') && customObject.result[0].dedupeFields.length > 0) {
+        const missingDedupeFields = [];
+        customObject.result[0].dedupeFields.forEach((field) => {
+          if (!Object.keys(object).some(e => e == field)) {
+            missingDedupeFields.push(field);
+          }
+        });
+
+        // Missing dedupe fields validation
+        if (missingDedupeFields.length > 0) {
+          return this.error('Error creating or updating %s: you must provide values for the the following fields: %s', [
+            name,
+            missingDedupeFields.join(','),
```

Comment:
- [ ] nit: recommend joining on `, ` (with space) to appear friendlier in context.

If we know this, it might be nice to include the field labels here in addition to the field names.  For example:

```
Error creating or updating fooBarCustomObject_c: you must provide values for the following fields: First Field (firstField), Another Field Label (anotherField)
```